### ISSUE TITLE
linkerd fails to route correctly for a recreated k8s namespace

### ISSUE THREAD

**Issue by agunnerson-ibm** [linkerd/linkerd#1170]
Description
---
When linkerd is set up to route to Kubernetes services, it seems to cache something related to the service such that if I delete and recreate the namespace, it can no longer route to services in that namespace.

```
No hosts are available for /http/1.1/GET/nginx, Dtab.base=[/k8s=>/#/io.l5d.k8s;/ns=>/k8s/master;/ns=>/k8s/production;/srv=>/ns/http;/http/*/*=>/srv], Dtab.local=[/ns=>/k8s/temp]. Remote Info: Not Available
```

The same thing seems to happen if I delete everything in the namespace and redeploy the services.

Environment
---
* Linkerd configuration (running as DaemonSet): https://gist.github.com/agunnerson-ibm/7884bf73964e12af93c45ad4ca1d758b#file-l5d-config-yaml
* Kubernetes 1.5.3 infrastructure and kubelet 1.5.3 sidecar container
* Kubernetes service and deployment: https://gist.github.com/agunnerson-ibm/7884bf73964e12af93c45ad4ca1d758b#file-nginx-yaml

Reproduction steps
---

I'm assuming that linkerd is set up similar to the configuration above.

Given the dtab in the linkerd config:

```
/k8s      => /#/io.l5d.k8s;
/ns       => /k8s/master;
/ns       => /k8s/production;
/srv      => /ns/http;
/http/*/* => /srv;
```

1. Deploy `nginx.yaml` to `temp` namespace:

```sh
kubectl -n temp apply -f nginx.yaml
```

2. Hit the service to make sure it works:

```sh
http_proxy=http://kube-node-01:4140 curl -v http://nginx -H 'l5d-dtab: /ns => /k8s/temp'
```

3. Delete the namespace and recreate it:

```sh
kubectl delete ns temp
# Wait for deletion to complete (see `kubectl get ns`)
kubectl create ns temp
```

4. Redeploy `nginx.yaml` to recreated `temp` namespace:

```sh
kubectl -n temp apply -f nginx.yaml
```

5. Linkerd can no longer route to the `temp` namespace:

```
$ http_proxy=http://kube-node-01:4140 curl http://nginx -H 'l5d-dtab: /ns => /k8s/temp'
No hosts are available for /http/1.1/GET/nginx, Dtab.base=[/k8s=>/#/io.l5d.k8s;/ns=>/k8s/master;/ns=>/k8s/production;/srv=>/ns/http;/http/*/*=>/srv], Dtab.local=[/ns=>/k8s/temp]. Remote Info: Not Available
```

**Comment by agunnerson-ibm** [comment]
I don't know if it's related, but I found that if you delete a kubernetes service, but keep the deployment/pods around, then linkerd still routes to those pods (instead of returning some sort of error).

**Comment by adleong** [comment]
@agunnerson-ibm that behavior is intentional.  we've found that treating service discovery as advisory allows us to be more resilient to service discovery outages.  The behavior can be configured with the `enableProbation` flag: https://linkerd.io/config/0.9.1/linkerd/index.html#load-balancer

**Comment by wmorgan** [comment]
@agunnerson-ibm thanks for the report with detailed repro instructions! Very helpful. Namespace thing is probably a bug. We'll look into it. Probation thing, like Alex says, is working as intended.

**Comment by agunnerson-ibm** [comment]
> @agunnerson-ibm that behavior is intentional. we've found that treating service discovery as advisory allows us to be more resilient to service discovery outages. The behavior can be configured with the enableProbation flag: https://linkerd.io/config/0.9.1/linkerd/index.html#load-balancer

> @agunnerson-ibm thanks for the report with detailed repro instructions! Very helpful. Namespace thing is probably a bug. We'll look into it. Probation thing, like Alex says, is working as intended.

Understood. Thanks!

**Comment by manishrajkarnikar** [comment]
@adleong I set enableProbation to true. Then deleted my service and recreated it. It could not route to the new service

**Comment by nik786** [comment]
linkerd conatiners are not getting created on kubernetes 1.5. Is Linkerd compatible with kubernetes 1.5 ??I have executed below mentioned command to install it with kubernetes
kubectl apply -f https://raw.githubusercontent.com/linkerd/linkerd-examples/master/k8s-daemonset/k8s/linkerd.yml...Containers are in creating state or crash loopback state

**Comment by adleong** [comment]
@nik786 it doesn't seem like that has anything to do with this issue.  Posting on discourse.linkerd.io is the best way to get help debugging your setup.

**Comment by agunnerson-ibm** [comment]
It looks like this bug was finally fixed in linkerd 1.2.1! I was able reproduce it in 1.2.0, but not 1.2.1.

**Comment by wmorgan** [comment]
@agunnerson-ibm awesome, glad to hear it!
